File: BE/src/model/vocabulary.model.ts
import mongoose, { Document, Schema } from "mongoose";

export interface IVocabulary {
  word: string;
  reading: string;
  meaning: string;
  level: "N1" | "N2" | "N3" | "N4" | "N5";
  topic: string;
  example?: string;
  exampleMeaning?: string;
  tags?: string[];
  createdAt?: Date;
  updatedAt?: Date;
}

export interface IVocabularyDocument extends IVocabulary, Document {}

const vocabularySchema = new Schema<IVocabularyDocument>(
  {
    word: { type: String, required: true, trim: true },
    reading: { type: String, required: true, trim: true },
    meaning: { type: String, required: true, trim: true },
    level: {
      type: String,
      enum: ["N1", "N2", "N3", "N4", "N5"],
      required: true,
    },
    topic: { type: String, required: true, trim: true },
    example: { type: String, trim: true },
    exampleMeaning: { type: String, trim: true },
    tags: [{ type: String, trim: true }],
  },
  { timestamps: true }
);

// Index for fast filtering
vocabularySchema.index({ level: 1, topic: 1 });
vocabularySchema.index({ word: "text", meaning: "text" });

export const Vocabulary = mongoose.model<IVocabularyDocument>(
  "Vocabulary",
  vocabularySchema
);

File: mirai_mobile/lib/models/vocabulary_model.dart
class VocabularyModel {
  final String id;
  final String word;
  final String reading;
  final String meaning;
  final String level;
  final String topic;
  final String? example;
  final String? exampleMeaning;
  final List<String>? tags;

  VocabularyModel({
    required this.id,
    required this.word,
    required this.reading,
    required this.meaning,
    required this.level,
    required this.topic,
    this.example,
    this.exampleMeaning,
    this.tags,
  });

  factory VocabularyModel.fromJson(Map<String, dynamic> json) {
    var tagList = json['tags'] as List? ?? [];
    List<String> parsedTags = tagList.map((t) => t.toString()).toList();

    return VocabularyModel(
      id: json['_id']?.toString() ?? json['id']?.toString() ?? '',
      word: json['word']?.toString() ?? '',
      reading: json['reading']?.toString() ?? '',
      meaning: json['meaning']?.toString() ?? '',
      level: json['level']?.toString() ?? 'N5',
      topic: json['topic']?.toString() ?? '',
      example: json['example']?.toString(),
      exampleMeaning: json['exampleMeaning']?.toString(),
      tags: parsedTags,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'word': word,
      'reading': reading,
      'meaning': meaning,
      'level': level,
      'topic': topic,
      'example': example,
      'exampleMeaning': exampleMeaning,
      'tags': tags,
    };
  }
}
